public/app/plugins/panel/geomap/components/MeasureVectorLayer.ts

```diff
@@ -0,0 +1,243 @@
+import VectorLayer from 'ol/layer/Vector';
+import VectorSource from 'ol/source/Vector';
+import { getArea, getLength } from 'ol/sphere';
+import { Circle as CircleStyle, Fill, RegularShape, Stroke, Style, Text } from 'ol/style';
+
+import { formattedValueToString } from '@grafana/data';
+
+import { MapMeasureOptions, measures } from '../utils/measure';
+
+export class MeasureVectorLayer extends VectorLayer<VectorSource> {
+  opts: MapMeasureOptions = {
+    action: 'length',
+    unit: 'm',
+  };
+  constructor() {
+    super({
+      source: new VectorSource(),
+    });
+    this.setStyle((feature) => {
+      return this.styleFunction(feature, false);
+    });
+    this.setVisible(true);
+  }
+
+  setOptions(options: MapMeasureOptions) {
+    this.opts = options;
+    this.getSource()?.refresh();
+  }
+
+  getMapMeasurement(geo: Geometry): string {
+    let v = 0;
+    let action = measures[0];
+    if (this.opts.action === 'area') {
+      action = measures[1];
+      v = getArea(geo);
+    } else {
+      v = getLength(geo);
+    }
+    return formattedValueToString(action.getUnit(this.opts.unit).format(v));
+  }
+
+  segmentStyle = new Style({
+    text: new Text({
+      font: '12px Calibri,sans-serif',
+      fill: new Fill({
+        color: 'rgba(255, 255, 255, 1)',
+      }),
+      backgroundFill: new Fill({
+        color: 'rgba(0, 0, 0, 0.4)',
+      }),
+      padding: [2, 2, 2, 2],
+      textBaseline: 'bottom',
+      offsetY: -12,
+    }),
+    image: new RegularShape({
+      radius: 6,
+      points: 3,
+      angle: Math.PI,
+      displacement: [0, 8],
+      fill: new Fill({
+        color: 'rgba(0, 0, 0, 0.4)',
+      }),
+    }),
+  });
+
+  segmentStyles = [this.segmentStyle];
+
+  // Open Layer styles
+  styleX = new Style({
+    fill: new Fill({
+      color: 'rgba(255, 255, 255, 0.2)',
+    }),
+    stroke: new Stroke({
+      color: 'rgba(0, 0, 0, 0.5)',
+      lineDash: [10, 10],
+      width: 2,
+    }),
+    image: new CircleStyle({
+      radius: 5,
+      stroke: new Stroke({
+        color: 'rgba(0, 0, 0, 0.7)',
+      }),
+      fill: new Fill({
+        color: 'rgba(255, 255, 255, 0.2)',
+      }),
+    }),
+  });
+
+  labelStyle = new Style({
+    text: new Text({
+      font: '14px Calibri,sans-serif',
+      fill: new Fill({
+        color: 'rgba(255, 255, 255, 1)',
+      }),
+      backgroundFill: new Fill({
+        color: 'rgba(0, 0, 0, 0.7)',
+      }),
+      padding: [3, 3, 3, 3],
+      textBaseline: 'bottom',
+      offsetY: -15,
+    }),
+    image: new RegularShape({
+      radius: 8,
+      points: 3,
+      angle: Math.PI,
+      displacement: [0, 10],
+      fill: new Fill({
+        color: 'rgba(0, 0, 0, 0.7)',
+      }),
+    }),
+  });
+
+  tipStyle = new Style({
+    text: new Text({
+      font: '12px Calibri,sans-serif',
+      fill: new Fill({
+        color: 'rgba(255, 255, 255, 1)',
+      }),
+      backgroundFill: new Fill({
+        color: 'rgba(0, 0, 0, 0.4)',
```

Comment:
Not sure on having these colors hardcoded to black, depending on the layer type this tool becomes less useful as user is unable to see lines being drawn :/ This can be a bug we generate to investigate a solution for and doesn't have to be a blocker on release - though for the default layer on dark mode this appears to be an issue so may be worth looking into for initial release

https://user-images.githubusercontent.com/22381771/182710492-85d8e1f9-1cb6-4884-9799-c015594393de.mov

Reply:
I have an idea for this. Let me see if I can implement quickly.

Reply:
![Aug-03-2022 15-15-33](https://user-images.githubusercontent.com/60050885/182721870-58e51a08-33b7-4e78-b44e-8e5767b8626d.gif)
![Aug-03-2022 15-15-12](https://user-images.githubusercontent.com/60050885/182721874-ea43d81e-6ef5-4a2b-93bd-2494fccb3477.gif)